### PR TITLE
df_parser parent_env unit tests

### PR DESCRIPTION
**Pull request by rcerven**
unit tests for #584

### CODE REVIEW

File: tests/test_util.py [hunk 1]
```diff
@@ -456,3 +459,46 @@ def test_df_parser(tmpdir):
     assert len(df.labels) == 1
     assert df.labels.get('label') == 'foobar barfoo'
 
+def test_df_parser_parent_env_arg(tmpdir):
+    tmpdir_path = str(tmpdir.realpath())
+    p_env = {
+        "test_env": "first"
+    }
+    df_content = """\
+FROM fedora
```
Comment (twaugh) [inline tests/test_util.py]: Could use textwrap's `dedent()` here (as in other places) to make it easier to read.

File: tests/test_util.py [hunk 2]
```diff
@@ -456,3 +459,46 @@ def test_df_parser(tmpdir):
...
+ENV foo=bar
+LABEL label="foobar $test_env"
+"""
+   
+    with open(tmpdir_path + '/Dockerfile', 'w') as dockerfile:
+        dockerfile.write(df_content)
+        dockerfile.flush()
+
+    df = df_parser(tmpdir_path, parent_env=p_env)
+    assert df.parent_env.get('test_env') == 'first'
```
Comment (twaugh) [inline tests/test_util.py]: I think `parent_env` isn't really meant to be a public part of the interface for that object.

File: tests/test_util.py [hunk 3]
```diff
@@ -456,3 +459,46 @@ def test_df_parser(tmpdir):
...
+   
+    with open(tmpdir_path + '/Dockerfile', 'w') as dockerfile:
+        dockerfile.write(df_content)
+        dockerfile.flush()
+
+    workflow = DockerBuildWorkflow(MOCK_SOURCE, 'test-image')
+    flexmock(workflow, base_image_inspect=env_conf)
+
+    df = df_parser(tmpdir_path, workflow=workflow)
+    assert df.parent_env.get('test_env') == 'first'
```
Comment (twaugh) [inline tests/test_util.py]: As above, not sure this line is necessary.

File: tests/test_util.py [hunk 4]
```diff
@@ -456,3 +483,57 @@ def test_df_parser(tmpdir):
...
+
+    def restore_logger():
+        util.logger = existing_logger
+    
+    request.addfinalizer(restore_logger)
+    df = df_parser(tmpdir_path, workflow=workflow)
+
+    if isinstance(env_arg, list) and ('=' not in env_arg[0]):
+        expected_log_message = ("Unable to parse all of Parent Config ENV")
+        assert (expected_log_message, ) in fake_logger.debugs
```
Comment (twaugh) [inline tests/test_util.py]: You can use `caplog` to do this sort of thing and it's generally easier. Just add 'caplog' to the parameter list for that method and the pytest caplog fixture will be passed in -- there are other examples around in this project.

You probably want something like:
```
with caplog.atLevel(logging.DEBUG):
   ...  # use df_parser()

assert msg in [x.message for x in caplog.records()]
```

File: tests/test_util.py [hunk 5]
```diff
@@ -456,3 +460,49 @@ def test_df_parser(tmpdir):
...
+    p_env = {
+        "test_env": "first"
+    }
+    df_content = dedent("""\
+        FROM fedora
+        ENV foo=bar
+        LABEL label="foobar $test_env"
+        """)
+
+    with open(tmpdir_path + '/Dockerfile', 'w') as dockerfile:
```
Comment (vrutkovs) [inline tests/test_util.py]: We don't really need that:
```
    df = df_parser(str(tmpdir))
    df.content = df_content
```
should be sufficient

File: tests/test_util.py [hunk 6]
```diff
@@ -456,3 +460,49 @@ def test_df_parser(tmpdir):
...
+
+    df = df_parser(tmpdir_path, parent_env=p_env)
+    assert df.labels.get('label') == 'foobar first'
+
+@pytest.mark.parametrize('env_arg', [
+    {"test_env": "first"},
+    ['test_env=first'],
+    ['test_env_first'],
+])
+def test_df_parser_parent_env_wf(tmpdir, caplog, env_arg):
```
Comment (vrutkovs) [inline tests/test_util.py]: Those testcases should probably be joined to test 'parent image envs + inspected config'

File: tests/test_util.py [hunk 7]
```diff
@@ -456,3 +460,49 @@ def test_df_parser(tmpdir):
...
+    with open(tmpdir_path + '/Dockerfile', 'w') as dockerfile:
+        dockerfile.write(df_content)
+        dockerfile.flush()
+
+    df = df_parser(tmpdir_path, parent_env=p_env)
+    assert df.labels.get('label') == 'foobar first'
+
+@pytest.mark.parametrize('env_arg', [
+    {"test_env": "first"},
+    ['test_env=first'],
```
Comment (vrutkovs) [inline tests/test_util.py]: Why would env vars lib contain lists?

Also, in `pre_distribution_scope` `base_image_inspect` contains a dict 'Labels' - is this test correct?

**Comment by vrutkovs** [comment]
Code added in https://github.com/projectatomic/atomic-reactor/pull/586

File: tests/test_util.py (hunk 7 above)
Comment (vrutkovs) [inline tests/test_util.py]: See the actual inspect output: it has to be a list of dicts (not a dict like in a first option and not a list of strings like in the last option). 

Instead it should test that 
* an empty list is parsed correctly (no envs in the base image), 
* a list of strings separated by '='

**Pull request merged**